src/app/zones/features.ts

```diff
@@ -0,0 +1,12 @@
+import type { Can, Features } from '@/app/application/services/can'
+import type Env from '@/services/env/Env'
```

Comment:
**Suggestion**: I think we should make this a more type-safe so it’s not possible to introduce `can` checks against features we don’t know about/don’t define (e.g. through typos).

This can be accomplished like that (though it would be nicer to derive the feature keys from the features.ts files; regardless it’s an improvement):

```suggestion
type FeatureKey =
  'create zones' |
  'use zones'
export type Can = (str: FeatureKey, obj?: any) => boolean
export type Features = Record<FeatureKey, (can: Can, obj?: any) => boolean>
```

Reply:
> it’s not possible to introduce can checks against features we don’t know about/don’t define (e.g. through typos)

This is a feature not a bug (the unknown features not the typo bit). i.e. you should be able to introduce features we don't know about/don't define, and these default to `false` i.e. the thing you are guarding doesn't show.

Whilst I agree it would be nice to protect against typos I don't think your suggestion will work with the composable/modular nature of this (the type definition you've suggested here I don't think applies to the `can` function that we use?), so I'm happy to leave that as something we can figure a solution for later. That being said I am looking at a potential solution for this as we speak, but not saying that will happen in this PR. Maybe think of these identifiers like route names in the meantime which are also prone to typos

All in all if you use `can` I'm sure you'd check what you have written works and/or maybe provide a test for it?

Reply:
> the type definition you've suggested here I don't think applies to the `can` function that we use?

I tried it and verified it’s working locally so I would think it does apply.

> This is a feature not a bug (the unknown features not the typo bit). i.e. you should be able to introduce features we don't know about/don't define, and these default to false i.e. the thing you are guarding doesn't show.

The unknown features part of your comment is handled by this suggestion as well: You can update the type and implement a feature function that just returns `true`.

> That being said I am looking at a potential solution for this as we speak, but not saying that will happen in this PR. Maybe think of these identifiers like route names in the meantime which are also prone to typos
> 
> All in all if you use `can` I'm sure you'd check what you have written works and/or maybe provide a test for it?

But for the `can` service, it’s easily done so we don’t even need to rely on us double-checking the spelling. This would be harder (as in a lot of work) for route names for example.

Reply:
> I tried it and verified it’s working locally so I would think it does apply.

Oh I see, sorry I was looking at the wrong file.

Ok so doing this here means we can't inject features from elsewhere, so its a no go.

> The unknown features part of your comment is handled by this suggestion as well: You can update the type and implement a feature function that just returns true.

Well yes you could, (returning false)

> it’s easily done

See above comment on injection, the can service shouldn't be aware of the individual modules.

Reply:
Alright then. I think it would be a significant improvement for the cost of not upholding the ideal that the keywords (mind you, they’re just a type, not JS values) are defined in their respective modules.

If you can think of a way to make this work, I think that would be super helpful.

Reply:
> the cost of not upholding the ideal

Its not an "ideal", its a necessity, we have other modules elsewhere as you know, hence we can't add them here. "Ideally" we'd have typo protection, but just make sure you check what you build.

> I think that would be super helpful.

Agree